### PR TITLE
Support Saving Opt Var Grad Tape

### PR DESCRIPTION
**Pull request by NihalHarish**
### Description of changes:
- Save optimizer variables for gradient tape
- Make call to `_write_optimizer_variables` in `_wrap_tape_gradient`
- Introduced a new function `_prepare_opt_variables` to prepare only optimizer variables since there is no call to `_prepare_non_layer_tensors`.
- Change assert statements to count optimizer variables

#### This PR also refactors code
Changes related to refactor:

- New fn `_get_custom_and_default_collections` that returns a tuple of two sets, `custom` and `default` collections
- The new fn is tested with `test_get_custom_and_default_collections` in `tests/tensorflow/hooks/test_collection_defaults.py`

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smdebug/tensorflow/keras.py [hunk 1]
```diff
@@ -805,6 +821,7 @@ def record_tensor_value(self, tensor_name, tensor_value):
             return
 
         self._add_metric(metric_name=tensor_name, metric_value=tensor_value)
+        self._write_optimizer_variables()
```
Comment (Vikas-kum) [inline smdebug/tensorflow/keras.py:824]: all the changes in this PR is same as https://github.com/awslabs/sagemaker-debugger/pull/218/files , 
Why this extra change isr eqd, I mean why do we have to write optimizer variables at 2 different places , here and in L573

Comment (NihalHarish) [reply smdebug/tensorflow/keras.py]: We make a call to `self._write_optimizer_variables()` because the `on_batch_begin` and `on_batch_end` callbacks are not called with GradientTape.

File: smdebug/tensorflow/keras.py [hunk 2]
```diff
@@ -739,6 +730,23 @@ def run(*args, **kwargs):
 
         return run
 
+    def _prepare_opt_variables(self):
+        optimizer_variables_collection = self.get_collection(CollectionKeys.OPTIMIZER_VARIABLES)
+        custom_collections, _ = self._get_custom_and_default_collections()
+
+        for tensor_ref in optimizer_variables_collection.get_tensors():
```
Comment (vandanavk) [inline smdebug/tensorflow/keras.py:737]: do we need tensor_ref or is opt variable available as an EagerTensor when gradtape is used?

File: tests/tensorflow2/test_keras.py [hunk 3]
```diff
@@ -139,6 +139,7 @@ def helper_keras_gradtape(
                     hook.get_collection(cname).save_config = SaveConfig(end_step=0)
 
     opt = tf.keras.optimizers.Adam()
+    hook.wrap_optimizer(opt)
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:142]: which case in wrap_optimizer is used when gradtape is used?

File: tests/tensorflow2/test_keras.py [hunk 4]
```diff
@@ -180,9 +181,10 @@ def test_keras_gradtape(out_dir, saveall):
 
     trial = smd.create_trial(path=out_dir)
     if saveall:  # save losses, metrics, weights, biases
-        assert len(trial.tensor_names()) == 10
+        assert len(trial.tensor_names()) == 15
         assert len(trial.tensor_names(collection=CollectionKeys.BIASES)) == 2
         assert len(trial.tensor_names(collection=CollectionKeys.WEIGHTS)) == 2
+        assert len(trial.tensor_names(collection=CollectionKeys.OPTIMIZER_VARIABLES)) == 5
```
Comment (vandanavk) [inline tests/tensorflow2/test_keras.py:187]: check vanilla TF as well

File: tests/tensorflow2/test_keras.py (hunk 3 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:142]: We need this to support vanilla TF

File: tests/tensorflow2/test_keras.py (hunk 4 above)
Comment (NihalHarish) [reply tests/tensorflow2/test_keras.py:187]: Works if we add `hook.wrap_optimizer(opt)`

File: smdebug/tensorflow/keras.py (hunk 2 above)
Comment (NihalHarish) [reply smdebug/tensorflow/keras.py]: I have removed the `prepare_opt_var` fn

File: smdebug/tensorflow/base_hook.py [hunk 5]
```diff
@@ -215,6 +217,18 @@ def export_collections(self):
         collection_file_name = f"{self.worker}_collections.json"
         self.collection_manager.export(self.out_dir, collection_file_name)
 
+    def _get_custom_and_default_collections(self) -> Tuple[Set["Collection"], Set["Collection"]]:
```
[the remark this answers is not in the document]
Comment (NihalHarish) [reply smdebug/tensorflow/base_hook.py:220]: Yes, since we currently do not update `custom` and `default` collections after training has started.
